### PR TITLE
New APIs for Schemas

### PR DESCRIPTION
**Pull request by aaronjeline**
## Description of changes
Adds a bunch of accessors for schemas:
* `principals()` gets all principal types
* `resources()` gets all resource types
* `principals_for_action()` get all principal types applicable to an action
* `resources_for_action()` get all resource types applicable to an action
* `parents()` get all the parent types of a type
* `action_groups()` get all action groups
* `enitty_types()` get all entity types
* `action_euids()` get all actions
## Issue #, if available
#1134 
## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [X] A backwards-compatible change requiring a minor version bump to `cedar-policy` (e.g., addition of a new API).

I confirm that this PR (choose one, and delete the other options):

- [x] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [X] Does not require updates because my change does not impact the Cedar formal model or DRT infrastructure.

### CODE REVIEW

File: cedar-policy/src/api.rs [hunk 1]
```diff
@@ -1428,6 +1428,79 @@ impl Schema {
     pub fn action_entities(&self) -> Result<Entities, EntitiesError> {
         Ok(Entities(self.0.action_entities()?))
     }
+
+    /// Returns an iterator over every entity type that can be a principal for any action in this schema
+    ///
+    /// Note: this iterator may contain duplicates.
+    pub fn principals(&self) -> impl Iterator<Item = &EntityTypeName> {
```
Comment (john-h-kastner-aws) [inline cedar-policy/src/api.rs:1435]: I feel like a lot of use cases for this function won't want duplicates, so we should make this easier by adding a doc-test showing this function used and collected into a hashset.

**Review by john-h-kastner-aws — APPROVED** [review]

File: cedar-policy-validator/src/schema.rs [hunk 2]
```diff
@@ -207,6 +207,86 @@ impl TryFrom<json_schema::Fragment<RawName>> for ValidatorSchema {
...
+            .get(action)
+            .map(ValidatorActionId::resources)
+    }
+
+    /// Returns an iterator over all the entity types that can be a parent of `ty`
+    ///
+    /// # Errors
+    ///
+    /// Returns [`None`] if the `ty` is not found in the schema
+    pub fn parents<'a>(
```
Comment (cdisselkoen) [inline cedar-policy-validator/src/schema.rs:257]: If I understand correctly from the implementation, this actually means (transitive) ancestors?  Should we name the method `ancestors()`?

File: cedar-policy-validator/src/schema.rs [hunk 3]
```diff
@@ -207,6 +207,86 @@ impl TryFrom<json_schema::Fragment<RawName>> for ValidatorSchema {
...
+            if action.descendants.is_empty() {
+                None
+            } else {
+                Some(&action.name)
+            }
+        })
+    }
+
+    /// Returns an iterator over all actions defined in this schema
+    pub fn action_euids(&self) -> impl Iterator<Item = &EntityUID> {
```
Comment (cdisselkoen) [inline cedar-policy-validator/src/schema.rs:286]: I suggest just `actions()`.  The above function wasn't named `action_group_euids()`, even though it also returns an iterator of `&EntityUID`

File: cedar-policy/CHANGELOG.md [hunk 4]
```diff
@@ -13,6 +13,8 @@ Starting with version 3.2.4, changes marked with a star (*) are _language breaki
 Cedar Language Version: 4.0
 
 ### Added
+- A bunch of new APIs for schemas. Allows you to query
+  principals,actions, and resources. (#1141, resolving #1134)
```
Comment (cdisselkoen) [inline cedar-policy/CHANGELOG.md:17]:
```suggestion
- A bunch of new APIs for schemas. Allows you to query
  principals, actions, and resources. (#1141, resolving #1134)
```

File: cedar-policy/src/api.rs [hunk 5]
```diff
@@ -1428,6 +1428,79 @@ impl Schema {
...
+    pub fn entity_types(&self) -> impl Iterator<Item = &EntityTypeName> {
+        self.0
+            .entity_types()
+            .map(|(name, _)| RefCast::ref_cast(name))
+    }
+
+    /// Returns an iterator over all actions defined in this schema
+    pub fn action_euids(&self) -> impl Iterator<Item = &EntityUid> {
+        self.0.action_euids().map(RefCast::ref_cast)
+    }
```
Comment (cdisselkoen) [inline cedar-policy/src/api.rs:1503]: Same comment on naming here

File: cedar-policy/src/api.rs [hunk 6]
```diff
@@ -1428,6 +1428,79 @@ impl Schema {
...
+            .resources_for_action(&action.0)
+            .map(|iter| iter.map(RefCast::ref_cast))
+    }
+
+    /// Returns an iterator over all the entity types that can be a parent of `ty`
+    ///
+    /// # Errors
+    ///
+    /// Returns [`None`] if the `ty` is not found in the schema
+    pub fn parents<'a>(
```
Comment (cdisselkoen) [inline cedar-policy/src/api.rs:1479]: Same comment on naming here

**Review by cdisselkoen — APPROVED** [review]

**Review by cdisselkoen — APPROVED** [review]

**Pull request merged**